getGeocode.js

```diff
@@ -0,0 +1,66 @@
+const dotenv = require("dotenv");
+//note: this only works locally; in Lambda we use environment variables set manually
+dotenv.config();
+
+const fetch = require("node-fetch");
+const { Client } = require("@googlemaps/google-maps-services-js");
+const { generateKey } = require("./data/dataDefaulter");
+
+const getGeocode = async (name, street, zip) => {
+    const address = `${name},MA,${street},${zip}`;
+    const client = new Client();
+    try {
+        const resp = await client.geocode({
+            params: {
+                address,
+                key: process.env.GOOGLE_API_KEY,
+            },
+        });
+        return resp.data;
+    } catch (e) {
+        console.error(e.response.data);
+    }
+};
+
+const sleep = (milliseconds) => {
+    return new Promise((resolve) => setTimeout(resolve, milliseconds));
+};
+
+const getAllCoordinates = async (locations, cachedResults) => {
+    const existingLocations = cachedResults.reduce((acc, location) => {
+        const { resolvedLocation, latitude, longitude } = location;
+        if (resolvedLocation && latitude && longitude) {
+            acc[generateKey(location)] = {
+                resolvedLocation,
+                latitude,
+                longitude,
+            };
+            return acc;
+        } else {
+            return acc;
+        }
+    }, {});
+
+    const coordinateData = await Promise.all(
+        locations.map(async (location) => {
+            const { name = "", street = "", zip = "" } = location;
+            const locationInd = generateKey(location);
+
+            if (existingLocations[locationInd]) {
+                return { ...location, ...existingLocations[locationInd] };
+            } else {
+                const locationData = await getGeocode(name, street, zip);
+                await sleep(100);
```

Comment:
Can you tell me more about the rate limit? Is there a way to keep a count before we take a break? I especially wonder for if we have an instance where we only need to get X geolocations and X requests can be successfully transmitted rapidfire.

Reply:
From [Google API Docs](https://developers.google.com/maps/premium/usage-limits#maps-javascript-api)

** When you first load the API, you are allocated an initial quota of requests. Once you use this quota, the API enforces rate limits on additional requests on a per-second basis. If too many requests are made within a certain time period, the API returns an OVER_QUERY_LIMIT response code. The per-session rate limit prevents the use of client-side services for batch requests, such as batch geocoding. For batch requests, use our web service APIs.

Not sure what the initial quota is. Further down the page says 50 requests per second.

Reply:
Is there a way we can catch that response and only wait a second if we get it?